<a href="https://colab.research.google.com/github/adeeljames/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/adeeljames/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

Lane: Refresh / Content Opportunity Scoring (Lane 2)

I'm choosing this lane because it directly builds on what I already explored in Week 1. The starter pipeline's own results (Precision@50 = 0.740 for random forest vs 0.240 for the hand-written baseline rule) show that a learned model finds real, non-obvious signal in this exact problem. Content teams work with limited review capacity, so a ranked "review this first" queue is a concrete, useful deliverable — not just an academic exercise.

In [4]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

Unit of analysis: one page (one row = one content item, grouped by client).

Decision this improves: which pages a content reviewer opens first this week/sprint.

Action someone takes: a content strategist refreshes, expands, or updates the top-ranked pages.

Cost of a wrong call:
- False positive: wastes a reviewer's limited time on a page that wasn't really the highest priority.
- False negative: a truly declining page goes unnoticed, and the traffic/visibility loss keeps compounding.

Why data/ML can help at all: the starter pipeline shows a hand-written rule alone gets only ~24% of its top-50 picks right (Precision@50 = 0.240), while a random forest gets ~74% right (0.740) — roughly a 3x lift. That gap suggests there is real, learnable signal beyond simple heuristics.

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


In [8]:
import os
print("Current directory:", os.getcwd())
print(os.listdir("."))

Current directory: /content
['.config', 'sample_data']


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

Loading the starter dataset to check how much of the inventory is actually in scope for this lane — i.e. how many pages are currently declining, and how many are stale but still getting traffic.

In [11]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/adeeljames/flyrank-ml-internship"
REPO_DIR = "flyrank-ml-internship"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    if os.path.exists("requirements.txt"):
        subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)

print("Working dir:", os.getcwd())

Working dir: /content/flyrank-ml-internship


In [12]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

declining_rate = df["trend_direction"].str.lower().eq("down").mean()
stale_visible = ((df["days_since_last_update"] >= 180) & (df["impressions_90d"] >= 500)).mean()
median_impressions = df["impressions_90d"].median()

print(f"Declining rate (trend_direction == 'down'): {declining_rate:.3f}")
print(f"Share of pages stale AND visible (>=180 days, >=500 impressions): {stale_visible:.3f}")
print(f"Median impressions_90d across all pages: {median_impressions:.0f}")

Declining rate (trend_direction == 'down'): 0.542
Share of pages stale AND visible (>=180 days, >=500 impressions): 0.001
Median impressions_90d across all pages: 731


Observed: 54.2% of pages in this sample are currently in a "down" trend — a substantial
portion of the inventory, so a prioritization tool has real, ongoing work to do.

Only 0.1% of pages are both stale (180+ days since update) AND highly visible (500+
impressions/90d) — this combination is rare, which tells me the "stale AND visible"
rule alone is too narrow to be the primary signal; the lane will need broader or
combined signals (e.g. lower impression thresholds, or combining staleness with
declining trend) rather than relying on this single rule.

The median page gets 731 impressions over 90 days, showing there's meaningful search
visibility across the inventory worth protecting or improving.

This is direct evidence that: (a) there's a large enough declining population to
justify a ranking approach, and (b) simple single-condition rules like "stale AND
visible" are too strict on their own — reinforcing why a learned model (as shown in
the starter pipeline: Precision@50 = 0.740 vs baseline 0.240) adds real value over a
single hand-written threshold.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*


What I CAN claim:
- This is decision-support: it ranks pages by evidence so a human reviews the most promising ones first.
- Results describe observed patterns in this specific dataset, not proof of causation.
- The starter label (trend_direction == "down") is a proxy for decline based on the current window, not a validated future outcome — I will treat it as a starting point only.

What I CANNOT claim:
- I cannot claim a refresh causes a page to recover — that would require an actual experiment.
- I cannot claim to reproduce or explain Google's ranking algorithm.
- I will not use FlyRank product decision flags (health_score, priority_score, action_type) as model features or labels.

In [13]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.